### PR TITLE
Rest middleware refactoring

### PR DESCRIPTION
**Pull request by graetzer**
* Remove `RestEngine`
* Consolidated some `CommTask` logic into `GeneralCommTask`
  * introduce `prepareExecution` and `finishExecution` in `GeneralCommTask`
* Delete all WorkStack related code

http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr-linux/464/

### CODE REVIEW

File: arangod/V8Server/v8-vocbase.cpp [hunk 1]
```diff
@@ -2014,13 +2014,13 @@ void JS_ArangoDBContext(v8::FunctionCallbackInfo<v8::Value> const& args) {
...
-  if (context != nullptr) {
+  
+  ExecContext const* exec = ExecContext::CURRENT;
+  if (exec != nullptr) {
     result->Set(TRI_V8_ASCII_STRING(isolate, "user"),
-                TRI_V8_STD_STRING(isolate, context->_user));
-    result->Set(TRI_V8_ASCII_STRING(isolate, "database"),
-                TRI_V8_STD_STRING(isolate, context->_database));
+                TRI_V8_STD_STRING(isolate, exec->user()));
+    /*result->Set(TRI_V8_ASCII_STRING(isolate, "database"),
```
Comment (jsteemann) [inline arangod/V8Server/v8-vocbase.cpp]: remove if not needed

File: lib/Basics/Thread.cpp [hunk 2]
```diff
@@ -89,33 +88,21 @@ void Thread::startThread(void* arg) {
     TRI_SetProcessorAffinity(&ptr->_thread, ptr->_affinity);
   }
 
-  bool pushed = WorkMonitor::pushThread(ptr);
-
   try {
```
Comment (jsteemann) [inline lib/Basics/Thread.cpp]: can be simplified by using `TRI_DEFER(ptr->cleanupMe());` here.

File: lib/Basics/Thread.h [hunk 3]
```diff
@@ -108,29 +107,8 @@ class Thread {
...
-    if (description == nullptr) {
-      return nullptr;
-    }
-
-    return description->_context.get();
-  }
-
   // checks if work has been canceled
-  static bool isCanceled() {
+  /*static bool isCanceled() {
```
Comment (jsteemann) [inline lib/Basics/Thread.h]: remove if not needed.

File: lib/Basics/Thread.cpp (hunk 2 above)
Comment (graetzer) [reply lib/Basics/Thread.cpp]: why add another lambda to every thread ?

**Pull request merged**